### DISCUSSION TITLE
Use Tauri or Neutralino as a lightweight (non-bloat) alternative to electron

### DISCUSSION THREAD

**Discussion by yeus** [quasarframework/quasar#12509] [Ideas / Proposals]
Electron packages the complete node.js environemnt and a chrome browser. This makes SW built on electron quiet big. Some projects emerged that want to help out on this regard. 

Look at this blogpost:

https://blog.logrocket.com/why-use-electron-alternative/

or this one:

https://www.smashingmagazine.com/2020/07/tiny-desktop-apps-tauri-vuejs/

Wondering if there would be any issues using this with quasar? Electron does have quiet a bit of code-bloat :P

Any thoughts on this? I want to try it out and will report back here. But thought I would already put this out here to start a discussion.

**Comment by customautosys** [comment]
I hope this will be considered too.

**Comment by yeus** [comment]
I just want to report back that I followed the instruction https://tauri.app/v1/guides/getting-started/setup/integrate in order to setup tauri for my existing quasar project and it works great (that quasar project was built using `yarn create quasar`)!   really awesome in fact... Didn#t have any problems at all and tauri works out-of-the-box with the standard configuration with quasar.

There are some things that could be made more comfortable with quasar though:

- add quasar icon support for tauri
- integrate tauri API in quasar
- add a tauri flag 
- right now I have to launch `tauri dev` and `quasar dev` in parallel...  maybe that could be made more elegant as well..

more to come!

But my first impression is that tauri is a HUGE improvement over electron apps..  now trying to build a stand-alone app desktop app and will report the resulting size :P

EDIT: update links

**Reply by customautosys** [reply]
I'm currently waiting for AppX and Snap support.

Out of curiosity, do you happen to know if Tauri has crates to:
Convert HTML to PDF (yes I know about html2pdf and jspdf, but I have some issues with their PDF rendering - the built-in electron PDF print function works much better)
OCR PDFs (again I know about Tessaract JS - it is quite slow)

**Reply by customautosys** [reply]
Oh BTW I found this in case anyone is interested: https://github.com/harshkhandeparkar/tauri-snap-packager

**Reply by hermit99** [reply]
@yeus Was your existing quasar project using Electron and you migrated from electron to tauri?

**Reply by yeus** [reply]
@hermit99 no..  I started a fresh project ...

**Comment by kantum** [comment]
Tauri is very promising, however there are still things that don't work out of the box, for example [webrtc](https://github.com/tauri-apps/wry/issues/85)

**Comment by hermit99** [comment]
Now that Tauri reaches 1.0 milestone, just wondering what is the strategy from Quasar team to integrate with Tauri? Is there any plan or roadmap? Thanks

**Comment by lu-zero** [comment]
tauri is also about to gain mobile support, so it will be also an alternative to cordova.

**Comment by yeus** [comment]
just want to say I am happily using tauri in combination with quasar for a lot of things now haven#t run into any issues since then!.  :)

**Reply by lu-zero** [reply]
would you write a tutorial about it?

**Reply by maxiride** [reply]
Found a repo with Quasar + Tauri: https://github.com/ricardodarocha/vite-quasar-tauri-template

Didn't test its yet, it's not updated and might be unstable. Could be a starting point in case @yeus doesn't come back with some info on how he did it.

**Comment by fairking** [comment]
I work on some Wails + Quasar project. Also have no issues with that at all (so far).
Tauri? Rust is not my cup of tea. I feel it's too complicated in my opinion. I just don't want waste my time.

You can find Quasar template here: https://wails.io/docs/community/templates/#vue

**Comment by SymphonyCR** [comment]
Following this!, Tauri seems to be a game changer improvement over Electron, and if their mobile support works, it could be an all in one alternative to Capacitor/Cordova/Electron.

**Comment by Heartnett** [comment]


**Reply by yeus** [reply]
Not sure, but neutralino.js and tauri essentially follow the same concept, right? If you don't want to, you would never come into contact with any rust code in tauri (at least thats the case in my own project. Its pure js based I also run it as a PWA).  

They both use a lightweight javascript engine. And both of them offer a javascript API to do "local" thinks that don't work in a browser like interacting with the local file system. So except for the differences in their javascript API they both fulfill the same job. my issue neutralino is mostly that it depends on the javascript engine on your system. So when installing it in different systems, you potentially get different javascript engines and could break your app. Tauri avoids that by appending a lightweight rust engine.

**Reply by yeus** [reply]
> Here you are talking about your project. More often than not, there will be many instances where one will have to interact with > the file system, handle certain events, perform some heavy processing, interact with the OS, etc. For that you need to write rust.

No that's not correct .  Tauri provides a javascript API for all of the things you mentioned, you wouldn't need to touch any Rust code for that at all  (You will come in touch with the Rust package manager but nothing else from rust).  here is the documentation --> https://tauri.app/v1/api/js/

> I will not be considering tauri, unless a JS/TS API wrapper is developed for their rust engine.

well then I guess you can consider it now because that's exactly what they have already done a while ago ;-). Check the docs above.

I am not saying, that you should write Rust. And by using Tauri, with a very high likelyhood you will not come in touch with any rust code either. 

I agree that neutralino.js is an awesome project as well.  Tauri, Neutralino.js and Wails  are all super similar, follow the same concept. The small differences how I see them are:

- for neutralino, the build process is probably faster than tauri
- tauri provides its own web engine which guarantees your app works on all clients.
- tauri api is based on rust which probably makes it more secure on most systems.
- neutralino apps might be slightly smaller, as they integrate the OS's libraries. But we're talking about really small difference her.e..  a few MBs at most.
- neutralino is probably a little easier to configure in the beginning than tauri as there is no initial build process.

**Comment by Edge-coordinates** [comment]
Hey, guys, any new developments in this matter? Will there be any official support?

**Reply by customautosys** [reply]
Just wondering, has anyone tried SciterJS before? How was your experience compared to Neutralino and Tauri?

**Comment by NaturalDevCR** [comment]
Just following up this, it would be great

**Comment by mr-zwets** [comment]
It's a great time to re-visit this discussion as Tauri has recently had it's v2 stable release!

> A very much awaited part of this release is the mobile operating system support. The previous version of Tauri allowed to have a single UI code base for desktop operating systems but now this extends to iOS and Android.

As the website says 'Build your app for Linux, macOS, Windows, Android and iOS - all from a single codebase.'

'Minimal Size By using the OS’s native web renderer, the size of a Tauri app can be little as 600KB.'

https://v2.tauri.app/blog/tauri-20/

**Comment by mr-zwets** [comment]
I found it was incredibly easy to add `src-tauri` to an existing quasar project

I created a repo for it here to show the changes I did: https://github.com/mr-zwets/quasar-tauri

I simply copied the `src-tauri` from the tauri demo app into the folder of a demo `quasar-project` and then fixed the dependencies, scripts & configs 😄

**Reply by fairking** [reply]
I wish Wails supports Android. Cannot do rust, it's too harsh to me. 😄

**Reply by TobyMosque** [reply]
or even better, follow the docs and use the CLI:
```bash
pnpm create quasar
cd quasar_proj
pnpm add -D @tauri-apps/cli@latest
pnpm tauri init
pnpm tauri dev
```

**Reply by yeus** [reply]
> I wish Wails supports Android. Cannot do rust, it's too harsh to me. 😄

@fairking 

thats a misconception about tauri. You never have to come in contact with rust if you don't want to. We've had that discussion earlier in the thread but the post got deleted by the user (https://github.com/quasarframework/quasar/discussions/12509#discussioncomment-8390549). You can still check out my response. Tauri gives you a javascript API just like a browser or electron. Essentially, tauri is a very lightweight browser written in rust. But that doesn't mean have to *program* in rust. give it a go its really not hard to use :)

**Reply by zanzara** [reply]
@mr-zwets   Yep, Tauri sound's good so far.
But how about `electron-builder`  replacement then?
I have pretty much config stuff there, for `nsis`, `msi` ,`deb`, `snap` , `extraFiles` , `desktop`   ...
Is this all in Tauri yet configurable? 

...and btw: Also in `src-electron/electron-main.js`  I got pretty much "node.js" code in a project.
Do I would need to transfer all this code to rust then?  (aswell the interfaces for IPC in `electron-prelaad.js`  ?)

**Reply by yeus** [reply]
@zanzara 

tauri supports most installers out of the box:  (without any configuration it creates deb, Appimage, rpm) for me. Check out this list for all the supported installers, its huge, including all that you mentioned!

- https://tauri.app/distribute

> ...and btw: Also in src-electron/electron-main.js I got pretty much "node.js" code in a project.
Do I would need to transfer all this code to rust then? (aswell the interfaces for IPC in electron-prelaad.js ?)

Did you modify those files from quasar? If you didn't, you don't really need to do anything tauri does most of these things out-of-the-box. If you modified them, tauri gives you a "native" Api like controlling the window size, native http fetch without CORS restrictions, access to file system similar to nodejs:

- https://tauri.app/reference/javascript/api/#_top

Quasar doesn't prepare this for you like it does for electron. But its very easy to use this API and customize to your liking.

Alternatively if you need to run nodejs code e.g. a webserver, you can also add a nodejs binary to tauri with so-called "side-cars".

**Reply by zanzara** [reply]
@yeus   TY for reply.
But as I see, there are several install config options missing, particularly in `deb`  and `desktop` 

And yes, I had to added much stuff in `electron-main.js`   (ie.  from `crypto` ) , that I cannot find in this tauri api.

Idk what `side-cars` mean or what it is exactly, but if it binds a webserver and node.js,  then the benefit of having a smaller file distribution size with tauri against electron might be gone.
But this is only an assumption.

**Reply by fairking** [reply]
> @yeus 
> 
> thats a misconception about tauri. You never have to come in contact with rust if you don't want to.

Do you mean you can write golang in backend?

**Reply by yeus** [reply]
@fairking 

> Do you mean you can write golang in backend?

Not out-of-the-box. You would have to bundle it with a sidecar. I haven't done anything like this yet. But it seems to be a nice solution to because this means tauri would tare care of the bundling & distribution for you this way.
